readtags: regex pattern matching in expressions

masatake opened the pull request:
`#/patter/' can be used as an operator.

```
	    (`#/pattern/' string) => #t|#f
```

This notation is taken from Gauche.
`#/pattern/i' works case insensitive way.

An example session:
```
    $ cat tags
    !_TAG_FILE_FORMAT	2	/extended format; --format=1 will not append ;" to lines/
    ...
    !_TAG_PROGRAM_VERSION	0.0.0	/56b6bd51/
    ALPHA	f.py	/^class ALPHA:$/;"	c	language:Python	inherits:
    BETA	f.py	/^class BETA:$/;"	c	language:Python	inherits:
    GAMMA	f.py	/^class GAMMA (ALPHA, BETA):$/;"	c	language:Python	inherits:ALPHA, BETA

    $  ./readtags -Q '(#/PHA/ $inherits)' -ne -l
    GAMMA	f.py	/^class GAMMA (ALPHA, BETA):$/;"	kind:c	language:Python	inherits:ALPHA, BETA
```

masatake commented:
TOOD

- [ ] revise makefile 
- [ ] csv operator
- [ ] better way to report an error
- [ ] update man page
...

masatake commented:
@AmaiKinono, I have implemented regex pattern matching in readtags.
I would like you to try. For using from emacs, `string->regexp` is also added.
You can use the function like:
```
-Q '((string->regexp "A.*B") $name)'
```

Conceptually this one is converted while reading the expression:
```
-Q  '(#/A.*B/ $name)'
```

I would like you to try the new feature, and review the change for readtags.1 man page:
https://github.com/universal-ctags/ctags/pull/2551/commits/b2ea6ea1d9c60d537765889bd5407b9636adcc25

AmaiKinono commented:
Very cool!

For now I don't have any complaint of the regex engine itself, but:

- I didn't see the `#/PATTERN/` operator in `$ readtags -H filter`.

- How do we control the case sensitivity? Of course we could combine it with `downcase` to achieve this and that's totally fine, but I'm used to regex related program/functions to have the case sensitivity control, so I'm guessing we could do this here.

- Is the name `string->regexp` conventional in Scheme? I may be more used to `string-to-regexp`, or `regexp` is actually enough.

AmaiKinono commented on man/readtags.1.rst.in:
```diff
@@ -172,19 +173,34 @@ missing field is #f, so we could deal with missing fields:
   .. code-block:: console
 
-     $ readtags -Q '(member "A" $inherits)' -l
+     $ readtags -Q '(#/(^|,)A(,|$)/ $inherits)' -l
```
```suggestion
     $ readtags -Q '(#/(^|, )A(,|$)/ $inherits)' -l
```

You may need to refill the paragraphs after accepting my suggestions.

AmaiKinono commented on man/readtags.1.rst.in:
```diff
@@ -172,19 +173,34 @@ missing field is #f, so we could deal with missing fields:
-     $ readtags -Q '(member "A" $inherits)' -l
+     $ readtags -Q '(#/(^|,)A(,|$)/ $inherits)' -l
+
+Here ``$inherits`` holds a comma-separated class list like "A,B,C",
```
```suggestion
Here ``$inherits`` is a comma-separated class list like "A, B, C",
```

masatake replied:
I guess it never happens in the current implementation. However, I wonder we should fix the behavior of python parser... At least, as you wrote, the users must know the format inside inherits: are language-specific. I will write it to the man page or somewhere.

AmaiKinono commented on man/readtags.1.rst.in:
```diff
@@ -172,19 +173,34 @@ missing field is #f, so we could deal with missing fields:
+     $ readtags -Q '(#/(^|,)A(,|$)/ $inherits)' -l
+
+Here ``$inherits`` holds a comma-separated class list like "A,B,C",
+"Z,A", "P,A,Q", or just "A". A tags file read by readtags may have tag entries
```
```suggestion
"Z, A", "P, A, Q", or just "A". The tags file may have tag entries
```

AmaiKinono commented on man/readtags.1.rst.in:
```diff
@@ -172,19 +173,34 @@ missing field is #f, so we could deal with missing fields:
+
+Here ``$inherits`` holds a comma-separated class list like "A,B,C",
+"Z,A", "P,A,Q", or just "A". A tags file read by readtags may have tag entries
+that has no ``inherits:`` field. In that case ``$inherits`` holds #f,
```
```suggestion
that has no ``inherits:`` field. In that case ``$inherits`` is #f,
```

AmaiKinono commented on man/readtags.1.rst.in:
```diff
@@ -172,19 +173,34 @@ missing field is #f, so we could deal with missing fields:
+Here ``$inherits`` holds a comma-separated class list like "A,B,C",
+"Z,A", "P,A,Q", or just "A". A tags file read by readtags may have tag entries
+that has no ``inherits:`` field. In that case ``$inherits`` holds #f,
+and the regular expression matching raises an error; the regular
```
```suggestion
and the regular expression matching raises an error, since it
```

AmaiKinono commented on man/readtags.1.rst.in:
```diff
@@ -172,19 +173,34 @@ missing field is #f, so we could deal with missing fields:
+"Z,A", "P,A,Q", or just "A". A tags file read by readtags may have tag entries
+that has no ``inherits:`` field. In that case ``$inherits`` holds #f,
+and the regular expression matching raises an error; the regular
+expression matching works only for a string.
```
```suggestion
works only for strings. To avoid this problem:
```

AmaiKinono commented on man/readtags.1.rst.in:
```diff
@@ -172,19 +173,34 @@ missing field is #f, so we could deal with missing fields:
+
+  .. code-block:: console
+
+     $ readtags -Q '(and $inherits (#/(^|,)A(,|$)/ $inherits))' -l
```
```suggestion
     $ readtags -Q '(and $inherits (#/(^|, )A(,|$)/ $inherits))' -l
```

AmaiKinono commented on man/readtags.1.rst.in:
```diff
@@ -172,19 +173,34 @@ missing field is #f, so we could deal with missing fields:
-WARNING: It's planned to remove the ``member`` operator, and replace this
-technique with regular expression matching. By then, ``$inherits`` will not be
-a list, but the value of the ``inherits:`` field as it is.
+     $ readtags -Q '(and $inherits (#/(^|,)A(,|$)/i $inherits))' -l
```
```suggestion
     $ readtags -Q '(and $inherits (#/(^|, )A(,|$)/i $inherits))' -l
```

masatake commented:
@AmaiKinono, thank you for reviewing.
I will reflect the comments.

>  Is the name string->regexp conventional in Scheme? I may be more used to string-to-regexp, or regexp is actually enough

I took the name `string->regexp` from gauche, a scheme implementation I am familiar with.
Using `->` as part of the name of an operator for converting a type to another type is popular convetion in scheme.

masatake commented:
> I didn't see the #/PATTERN/ operator in $ readtags -H filter

I will implement it in another pull request.

masatake commented:
@AmaiKinono, I have reflected all your comments.
Could you look at 28f9d36 ?

masatake commented:
BTW, I found implementing macro expander is not difficult.

If you wrote following code in ~/.readtags
```
(defmacro is-method-for (class) (and (eq? $kind "method") (eq? $scope-kind "class") (eq? $scope-name class)))
```

You can use it like:
```
readtags -Q '(is-method-for "Window")' -l
```

AmaiKinono commented on docs/man/readtags.1.rst:
```diff
@@ -172,22 +173,42 @@ missing field is #f, so we could deal with missing fields:
+NOTE: The above regular expression pattern for inspecting inheritances is just an
+example to show how to use ``#/PATTERN/`` expression.  Tags file generators have
+no consensus about the format of ``inherits:``.  Even parsers in ctags have no
+consensus. Handling such language-specific aspects is inevitable in a client tool
```
Just a nitpicking: I've been careful when using the word "client tool" in this manpage. You can see it only appears once, and it's about using the `-E` option, i.e., about reading the output by a machine, not by human. But here, even interactive use needs to be aware of the format of the `inherits:` field, so I'd suggest change this sentence to:

```
Noticing the format of the ``inherits:`` field of specific languages is needed for such queries.
```

AmaiKinono commented on docs/man/readtags.1.rst:
```diff
@@ -172,22 +173,42 @@ missing field is #f, so we could deal with missing fields:
+no consensus about the format of ``inherits:``.  Even parsers in ctags have no
+consensus. Handling such language-specific aspects is inevitable in a client tool
+side.
+
```
Maybe we should talk a little more here:

```
Case-insensitive matching can be performed by ``#/PATTERN/i``. Readtags also
offers an alias ``string->regexp``, so ``#/PATTERN/`` is equal to
``(string->regexp "PATTERN")``, and ``#/PATTERN/i`` is equal to
``(string->regexp "PATTERN" :case-fold)``.
```

AmaiKinono commented:
About defmacro: I'm on the latest commit of your branch. I got this error when try your example:

```
COMPILE ERROR: unbound-variable: is-method-for
MEMORY EXHAUSTED or SYNTAX ERROR
Failed to compile the expression of filter: (is-method-for "A")
```

masatake commented:
Thank you very much. Updated.

I found I implemented string->regexp wrong way.
To do case-insenstive search, ``:case-fold #t`` or ``:case-fold true`` must be specified.
Just specifying ``:case-fold`` is not enough. Sorry for making confused. I already fix this.

About the macro, it is just implemtable IDEA. not implemented yet. If you think the macro feature is useful more than three times:-P, let me know.

AmaiKinono commented:
> To do case-insenstive search, `:case-fold #t` or `:case-fold true` must be specified.

I see.

> If you think the macro feature is useful more than three times:-P, let me know.

Actually I don't have a very positive opinion on it. Putting it in `~/.readtags` means it's mainly for interactive use, which I don't think people will do often. For a client tool, it can implement helpers that builds common sexp patterns, so it doesn't need readtags providing `defmacro`.

masatake commented:
> Actually I don't have a very positive opinion on it. Putting it in ~/.readtags means it's mainly for interactive use, which I don't think people will do often. For a client tool, it can implement helpers that builds common sexp patterns, so it doesn't need readtags providing defmacro.

I see. Thank you for the feedback.

masatake commented:
I would like to merge this tonight.

masatake commented:
@AmaiKinono, thank you for reviewing.
The next issue I have to fix is the exit status one.
